### PR TITLE
🗃 Feedback Flow

### LINKED ISSUE #256: Create a new table in database to collect feedback
With tests and all ❤️

### PR DESCRIPTION
**Pull request by astroash**
👀Please REVIEW me👀

**Disclaimer:**  I've been staring at this feature for too long so 🤷🏽‍♀️

**Breakdown**
- Attempt some architecture nonsense
- Update db files to include creating/destruction of new feedback table
- Create new route for /api/cb/feedback
- Create generalise `SELECT` query to take date ranges
- Create `INSERT` query for feedback magic
- Add feedback buttons to frontend
- Add thank you component with redirect magic
- 👷‍♂️ ALL OF THE TESTS 👷
- 📝 And even some docs

#255 
Fixes #256

### CODE REVIEW

**Comment by astroash** [comment]
@eliascodes is this what you were imagining with your [architecture plan](https://github.com/TwinePlatform/DataPower/blob/master/docs/architecture.md)?

File: db/schema.sql [hunk 1]
```diff
@@ -44,10 +44,18 @@ CREATE TABLE activities
...
   activitiesId INTEGER REFERENCES activities(id),
   cb_id INTEGER REFERENCES cbusiness(id),
   date TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT CURRENT_TIMESTAMP
 );
 
+CREATE TABLE feedback
+(
+  id SERIAL PRIMARY KEY,
+  cb_id INTEGER REFERENCES cbusiness(id),
+  user_feedback INTEGER NOT NULL,
```
Comment (eliascodes) [inline db/schema.sql]: 🙏 please can we destroy as many references to "user" as possible, IMO it's nothing but confusing. IMO `feedback_score` is better because we are collecting quantitative feedback, not qualitative.

File: react-backend/shared/models/cb-admin.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+  getCbFeedback: async (dbConnection, cbId) => {
```
Comment (eliascodes) [inline react-backend/shared/models/cb-admin.js]: If you imagine that this object is describing the `CbAdmin` model, then putting `CB` in the method name seems redundant; there are no other kinds of feedback we can collect.

```js
const feedback = await CbAdmin.getFeedback(client, id);

// compared to ...

const feedback = await CbAdmin.getCbFeedback(client, id);
```

Separate note (with no action right now): I usually like to make these methods as general as possible to capture future use cases more easily. For example, at the moment, this query can only give us all the feedback for a CbAdmin. But what if we want to filter by date in future? Wouldn't it be nice to be able to do something like:

```js
const then = new Date(Date.now() - 1e10)
const now = new Date(Date.now())
const feedback = await CbAdmin.getFeedback(client, { id, since: then, until: now });
```

File: tests/shared/models/cb-admin.test.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
+const test = require('tape');
+const pg = require('pg');
+const { getConfig } = require('../../../config');
+const { refresh: refreshDB } = require('../../../db/scripts');
+const cbAdmin = require('../../../react-backend/shared/models/cb-admin');
+
+const config = getConfig(process.env.NODE_ENV);
+test.only('Db Query | cbAdmin.getCbFeedback ', async tape => {
```
Comment (eliascodes) [inline tests/shared/models/cb-admin.test.js]: watch out for this `only`

File: react-backend/shared/models/cb-admin.js (hunk 2 above)
Comment (astroash) [reply react-backend/shared/models/cb-admin.js]: Oooooo

File: db/schema.sql (hunk 1 above)
Comment (astroash) [reply db/schema.sql]: Yeah, makes sense 😊

File: tests/shared/models/cb-admin.test.js (hunk 3 above)
Comment (astroash) [reply tests/shared/models/cb-admin.test.js]: Come on, its a [WIP]...

**Comment by astroash** [comment]
> Yeah this is pretty much along the lines of what I was thinking.

👌

File: react-backend/shared/models/cb-admin.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+const cbAdmin = {
+  then: new Date('2000-01-01T00:00:00.000Z'),
+  now: new Date(Date.now()),
+
+  getFeedback: async (
+    dbConnection,
+    { cbId, since = cbAdmin.then, until = cbAdmin.now }
```
Comment (astroash) [inline react-backend/shared/models/cb-admin.js]: @eliascodes, the internet tells me I can't do this but it works so 🤷🏽‍♀️...

File: react-backend/client/src/components/home.js [hunk 5]
```diff
@@ -1,11 +1,29 @@
...
+      feedbackScore,
+    }),
+  });
+};
 const Home = () => (
   <div>
     <h1>
       Who are you?<br />
     </h1>
+    <button onClick={() => postFeedback(1, -1)}>-1</button>
```
Comment (eliascodes) [inline react-backend/client/src/components/home.js]: where are these id's coming from?

File: react-backend/shared/models/cb-admin.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
+const then = new Date('2000-01-01T00:00:00.000Z');
+const now = new Date(Date.now());
+
+module.exports = {
+  getFeedback: async (dbConnection, { cbId, since = then, until = now }) => {
+    const getFeedbackQuery = `SELECT feedback_score, feedback_date FROM feedback WHERE cb_id = $1 AND feedback_date BETWEEN $2 AND $3`;
```
Comment (eliascodes) [inline react-backend/shared/models/cb-admin.js]: Is there an endpoint where this query is exposed?

File: react-backend/shared/models/cb-admin.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    const query = await dbConnection.query(getFeedbackQuery, [
+      cbId,
+      since,
+      until,
+    ]);
+    return query.rows;
+  },
+
+  postFeedback: async (dbConnection, { cbEmail, feedbackScore }) => {
```
Comment (eliascodes) [inline react-backend/shared/models/cb-admin.js]: Get rid of `post` here. The DB layer is not related to the API layer. There's no reason to introduce a coupling in the naming of methods and the endpoints on which they're used. My use of `get` in `getFeedback` was not a reference to the HTTP `GET` method. 

Instead, name the method for what it does in the context of the model. It adds a new feedback record. What might be a better name?

File: react-backend/shared/models/cb-admin.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        WHERE cbusiness.email = $1 LIMIT 1)
+    , $2) 
+    RETURNING *`;
+
+    const query = await dbConnection.query(postFeedbackQuery, [
+      cbEmail,
+      feedbackScore,
+    ]);
+
+    return query.rows;
```
Comment (eliascodes) [inline react-backend/shared/models/cb-admin.js]: If this query is returning only one record, why does it return an array? Remember #244

File: react-backend/client/src/components/home.js [hunk 9]
```diff
@@ -1,11 +1,29 @@
...
+  });
+
+  return fetch('/api/cb/feedback', {
+    method: 'POST',
+    headers,
+    body: JSON.stringify({
+      cbId,
+      feedbackScore,
+    }),
+  });
```
Comment (eliascodes) [inline react-backend/client/src/components/home.js]: Then what? How does the visitor know that their feedback was successfully or unsuccessfully recorded?

File: tests/shared/controllers/feedback.test.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+      const token = jwt.sign({ email: 'fakey@mcfake.com' }, secret);
+      const badPayload = {
+        feedbackScore: -1,
+      };
+
+      request(app)
+        .post('/api/cb/feedback')
+        .set('authorization', token)
+        .send(badPayload)
+        .expect(500)
```
Comment (eliascodes) [inline tests/shared/controllers/feedback.test.js]: Why does a bad payload result in a `5xx`? `4xx` errors denote problems with the request.

File: tests/shared/models/cb-admin.test.js [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+          feedback_date: new Date('2017-12-09T17:12:00.000Z'),
+        },
+      ];
+
+      t.deepEquals(
+        actualAll,
+        expectedAll,
+        'cbAdmin.getFeedback returns all cbs feedback without dates specified'
+      );
+      const sinceDate = new Date('2017-11-09 17:23:00+00');
```
Comment (eliascodes) [inline tests/shared/models/cb-admin.test.js]: Split these with/without dates into separate test cases.

File: docs/api_migration_plan.md [hunk 12]
```diff
@@ -82,3 +82,4 @@ All non-successful responses should return appropriate status codes in the range
 | `POST   /cbs/login` | | `{ query: { email, password } }` | `{ result: { cb_admin_token } }` | `POST /admin/login` |
 | `POST   /cbs/:id/emails` | | `{ query: { password_reset: Bool } }` | `{ result: null }` | `POST /cb/pwd/reset` |
 | `POST   /visitors/statistics` | `Authorization: cb_admin_token` | `{ query: { group_by: { gender, activity, age: [18, 35, 51, 70], days: ['monday'] } }` | `{ result: { gender: { male, female, ... }, activity: { yoga: 12, ...}, age: { 0-17: 2, ... }, days: { monday: 2, ... } } }` | `POST /users/chart-all` <br/> `POST /users/filtered` |
+| `POST   /cb/feedback` | `Authorization: cb_admin_token` | ` { email }` | `[ { cbEmail, feedbackScore, feedback_date } ]`| ✨ shiney shiney new ✨ |
```
[the remark this answers is not in the document]
Comment (astroash) [reply docs/api_migration_plan.md]: It depends how shiny it is....

File: react-backend/shared/controllers/feedback.js [hunk 13]
```diff
@@ -0,0 +1,21 @@
+const jwt = require('jsonwebtoken');
+const cbAdmin = require('../models/cb-admin');
+
+const route = ('/',
```
[the remark this answers is not in the document]
Comment (astroash) [reply react-backend/shared/controllers/feedback.js]: A function...

Your `x` is just a bunch of numbers wrapped in parens...

[the remark this answers is not in the document]
Comment (eliascodes) [reply react-backend/shared/controllers/feedback.js]: The value of `x` in my example is `3`. I was pointing out that assigning `(x, y, z)` to a variable is redundant, because `x` and `y` are ignored. You are assigning `('/', () => {...})` to `route`.

File: react-backend/shared/models/cb-admin.js [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+  },
+
+  postFeedback: async (dbConnection, { cbEmail, feedbackScore }) => {
+    const postFeedbackQuery = `
+    INSERT INTO feedback 
+    (cb_id, feedback_score) 
+    VALUES (
+      (SELECT cbusiness.id FROM cbusiness 
+        WHERE cbusiness.email = $1 LIMIT 1)
```
[the remark this answers is not in the document]
Comment (astroash) [reply react-backend/shared/models/cb-admin.js]: Na, I think this is left overs of a bizarre path I stumbled down trying to inner join on feedback n only match with emails of previously entries 😅

File: react-backend/shared/models/cb-admin.js [hunk 15]
```diff
@@ -0,0 +1,32 @@
+const then = new Date('2000-01-01T00:00:00.000Z');
+const now = new Date(Date.now());
```
[the remark this answers is not in the document]
Comment (astroash) [reply react-backend/shared/models/cb-admin.js]: Oops, yeah, I guess setting it to a function returning `Date.now()` instead?

File: react-backend/client/src/components/home.js (hunk 5 above)
Comment (astroash) [reply react-backend/client/src/components/home.js]: Yup, that explains the overwhelming feeling that I forgot something 😊

File: react-backend/shared/models/cb-admin.js (hunk 6 above)
Comment (astroash) [reply react-backend/shared/models/cb-admin.js]: Not yet.

File: react-backend/client/src/components/home.js (hunk 9 above)
Comment (RogeredBacon) [reply react-backend/client/src/components/home.js]: Are we gonna restrict how often feedback can be given? Some sort of waiting period, to stop people spamming. Maybe an issue for another time.

File: react-backend/shared/models/cb-admin.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (RogeredBacon) [reply react-backend/shared/models/cb-admin.js]: Good for safety but our cb_signup validation should do this anyway

File: react-backend/client/src/components/home.js (hunk 9 above)
Comment (astroash) [reply react-backend/client/src/components/home.js]: ☝️ Yeah, that was kinda my thinking. _Vaguely_ css related...

Comment (eliascodes) [reply react-backend/client/src/components/home.js]: Great shout @RogeredBacon. A frontend-only approach to abusive/malicious use cases is not a solution. Needs additional backend logic

File: react-backend/shared/controllers/feedback.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (astroash) [reply react-backend/shared/controllers/feedback.js]: Lol, oh yeah, I didn't realise I was doing that.

File: react-backend/shared/models/cb-admin.js [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+module.exports = {
+  getFeedback: async (dbConnection, { cbId, since = then, until = now }) => {
```
[the remark this answers is not in the document]
Comment (astroash) [reply react-backend/shared/models/cb-admin.js]: ## ☝️Option 3 ☝️ 
This was the **woes** I was talking about yesterday. I do not have access to the cb id in the front end where this query is called from, just the cb's email. I wanted to store the `cb_id` in the `feedback` table cos it's a foreign key, so I'm grabbing the data with the id, but have to insert & match on email address.

File: react-backend/shared/models/cb-admin.js (hunk 7 above)
Comment (astroash) [reply react-backend/shared/models/cb-admin.js]: Lol, I cannot help but read this in a primary school teacher voice...

File: react-backend/client/src/components/home.js (hunk 9 above)
Comment (astroash) [reply react-backend/client/src/components/home.js]: Oops, okay, I should have looked at the designs better, this is actually all on the wrong page and should have a redirect on the client... brb..

File: tests/shared/models/cb-admin.test.js (hunk 11 above)
Comment (astroash) [reply tests/shared/models/cb-admin.test.js]: How come? I was lumping them together under the idea that they were all testing different edge cases of the same function.

Comment (eliascodes) [reply tests/shared/models/cb-admin.test.js]: I guess it's a preference thing, but IMO it's hard to read and maintain long test cases.

File: react-backend/shared/controllers/feedback.js [hunk 17]
```diff
@@ -1,21 +1,24 @@
...
-const route = ('/',
-(req, res, next) => {
+module.exports = (req, res, next) => {
   const pgPool = req.app.get('client:psql');
   const standardJwtSecret = req.app.get('cfg').session.standard_jwt_secret;
   const { feedbackScore } = req.body;
   const { authorization: token } = req.headers;
 
   jwt.verify(token, standardJwtSecret, (err, decoded) => {
+    if (err) return next;
```
Comment (eliascodes) [inline react-backend/shared/controllers/feedback.js]: Forgot a bit

File: tests/shared/controllers/feedback.test.js [hunk 18]
```diff
@@ -43,9 +43,16 @@ test('POST /api/cb/feedback', tape => {
         .post('/api/cb/feedback')
         .set('authorization', token)
         .send(badPayload)
-        .expect(500)
-        .expect('Content-Type', /text/)
-        .end(async err => {
+        .expect(400)
```
Comment (eliascodes) [inline tests/shared/controllers/feedback.test.js]: Looking a bit closer at this test, the reason the request fails is not because the payload is bad. Actually there's nothing wrong with the payload at all. It's that the email in the token is non-existent. 

This means that the request is well formed, but unauthorized. That's a `401`, and the router should check whether the email exists first and respond with an appropriate message.

Why aren't you using the authentication middleware on this route? It would do all this for you.

Comment (astroash) [reply tests/shared/controllers/feedback.test.js]: Ahhh yeah, I should be using middleware 🙈 . Do you think I need an error test case, as this was checking the if the email grabbed from the jwt was linked to a valid cb. If it's running through middleware I guess this can never happen.

File: react-backend/shared/controllers/feedback.js [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+  jwt.verify(token, standardJwtSecret, (err, decoded) => {
+    if (err) return next;
+
+    const { email: cbEmail } = decoded;
+
+    cbAdmin
+      .insertFeedback(pgPool, { cbEmail, feedbackScore })
+      .then(data => res.send({ result: data }))
+      .catch(() =>
+        res
```
Comment (eliascodes) [inline react-backend/shared/controllers/feedback.js]: This `catch` will catch _any_ error occurring during the promise chain. It should be used for errors that we cannot anticipate and/or mitigate in advance (like programmer errors).

We _can_ anticipate payload validation errors. You're not doing any validation in this controller. Then you can correctly respond with a `400` for payload errors, and leave this as a `500`.

File: tests/shared/models/cb-admin.test.js [hunk 20]
```diff
@@ -0,0 +1,142 @@
+const test = require('tape');
+const pg = require('pg');
+const { getConfig } = require('../../../config');
+const { refresh: refreshDB } = require('../../../db/scripts');
+const cbAdmin = require('../../../react-backend/shared/models/cb-admin');
+
+const config = getConfig(process.env.NODE_ENV);
+
+test('Db Query | cbAdmin ', async tape => {
```
Comment (eliascodes) [inline tests/shared/models/cb-admin.test.js]: `CbAdmin model` is a better test suite descriptor.

File: tests/shared/models/cb-admin.test.js [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+  const client = new pg.Client(config.psql);
+  await client.connect();
+
+  tape.test('cbAdmin.getFeedback | existing cb_id', async t => {
```
Comment (eliascodes) [inline tests/shared/models/cb-admin.test.js]: Since they're nested inside the `CbAdmin model` test suite, there's no need to prefix the method names with `CbAdmin`, since that is implicit.

**Comment by eliascodes** [comment]
@astroash please rebase your branch onto master.

And btw, you need to leave a blank line between the commit summary and the long-form commit message for it to be recognised properly by all git clients:

screen shot 2018-03-14 at 14 25 07

File: react-backend/shared/controllers/feedback.js [hunk 22]
```diff
@@ -0,0 +1,24 @@
+const jwt = require('jsonwebtoken');
+const cbAdmin = require('../models/cb-admin');
+
+module.exports = (req, res, next) => {
+  const pgPool = req.app.get('client:psql');
+  const standardJwtSecret = req.app.get('cfg').session.standard_jwt_secret;
+  const { feedbackScore } = req.body;
+  const { authorization: token } = req.headers;
+
+  jwt.verify(token, standardJwtSecret, (err, decoded) => {
```
[the remark this answers is not in the document]
Comment (astroash) [reply react-backend/shared/controllers/feedback.js]: It's the only way I could get info about which cb is logged in on the front end.

File: react-backend/client/src/components/visitors/thank_you_feedback.js [hunk 23]
```diff
@@ -0,0 +1,32 @@
+import React, { Component } from 'react';
```
[the remark this answers is not in the document]
Comment (astroash) [reply react-backend/client/src/components/visitors/thank_you_feedback.js]: Sweet, that makes sense, should I get it to use [`this.props.history.goBack()`](https://reacttraining.com/react-router/web/api/history) instead?

File: react-backend/client/src/assets/icons/admin/Delete.svg [hunk 24]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (astroash) [reply react-backend/client/src/assets/icons/admin/Delete.svg]: Cool, I've never heard of this before. Do I simply just run it though [this lovely site I found on google?](http://petercollingridge.appspot.com/svg-optimiser)

File: docs/api_migration_plan.md [hunk 25]
```diff
@@ -82,3 +82,4 @@ All non-successful responses should return appropriate status codes in the range
 | `POST   /cbs/login` | | `{ query: { email, password } }` | `{ result: { cb_admin_token } }` | `POST /admin/login` |
 | `POST   /cbs/:id/emails` | | `{ query: { password_reset: Bool } }` | `{ result: null }` | `POST /cb/pwd/reset` |
 | `POST   /visitors/statistics` | `Authorization: cb_admin_token` | `{ query: { group_by: { gender, activity, age: [18, 35, 51, 70], days: ['monday'] } }` | `{ result: { gender: { male, female, ... }, activity: { yoga: 12, ...}, age: { 0-17: 2, ... }, days: { monday: 2, ... } } }` | `POST /users/chart-all` <br/> `POST /users/filtered` |
+| `POST   /cb/feedback` | `Authorization: cb_admin_token` | ` { email }` | `[ { cbEmail, feedbackScore, feedback_date } ]`| ✨ shiny shiny new ✨ |
```
[the remark this answers is not in the document]
Comment (astroash) [reply docs/api_migration_plan.md]: ewwww conform...

File: react-backend/shared/controllers/feedback.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (eliascodes) [reply react-backend/shared/controllers/feedback.js]: Did you read the auth middleware?

File: react-backend/client/src/components/visitors/thank_you_feedback.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (eliascodes) [reply react-backend/client/src/components/visitors/thank_you_feedback.js]: Sure

File: react-backend/client/src/assets/icons/admin/Delete.svg (hunk 24 above)
[the remark this answers is not in the document]
Comment (eliascodes) [reply react-backend/client/src/assets/icons/admin/Delete.svg]: yeah

[the remark this answers is not in the document]
Comment (astroash) [reply react-backend/client/src/assets/icons/admin/Delete.svg]: Are these default settings good?

screen shot 2018-03-14 at 18 31 13

[the remark this answers is not in the document]
Comment (eliascodes) [reply react-backend/client/src/assets/icons/admin/Delete.svg]: No idea

[the remark this answers is not in the document]
Comment (astroash) [reply react-backend/client/src/assets/icons/admin/Delete.svg]: LGTM :shipit:

File: tests/shared/models/cb-admin.test.js (hunk 20 above)
Comment (astroash) [reply tests/shared/models/cb-admin.test.js]: I've named all my model tests like this to make it easier to read the tape output with the nested tests we're using

File: tests/shared/models/cb-admin.test.js (hunk 21 above)
Comment (astroash) [reply tests/shared/models/cb-admin.test.js]: 🤷🏽‍♀️ Dunno why but I find this easier to read/reason about. Would prefer to leave it as is if that's okay.

File: react-backend/client/src/components/visitors/thank_you_feedback.js [hunk 26]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+
+export default () => (
```
Comment (eliascodes) [inline react-backend/client/src/components/visitors/thank_you_feedback.js]: Lol, these thanks components are still almost the same, but I'll let it go at this stage.

File: react-backend/client/src/components/hoc/redirect_after_timeout.js [hunk 27]
```diff
@@ -0,0 +1,27 @@
+import React, { Component } from 'react';
+import { withRouter } from 'react-router-dom';
+import PropTypes from 'prop-types';
+
+const redirectAfterTimeout = (url, ms) => (Child) => {
```
Comment (eliascodes) [inline react-backend/client/src/components/hoc/redirect_after_timeout.js]: This is really good.

File: docs/api_migration_plan.md [hunk 28]
```diff
@@ -82,3 +82,4 @@ All non-successful responses should return appropriate status codes in the range
 | `POST   /cbs/login` | | `{ query: { email, password } }` | `{ result: { cb_admin_token } }` | `POST /admin/login` |
 | `POST   /cbs/:id/emails` | | `{ query: { password_reset: Bool } }` | `{ result: null }` | `POST /cb/pwd/reset` |
 | `POST   /visitors/statistics` | `Authorization: cb_admin_token` | `{ query: { group_by: { gender, activity, age: [18, 35, 51, 70], days: ['monday'] } }` | `{ result: { gender: { male, female, ... }, activity: { yoga: 12, ...}, age: { 0-17: 2, ... }, days: { monday: 2, ... } } }` | `POST /users/chart-all` <br/> `POST /users/filtered` |
+| `POST   /cb/feedback` | `Authorization: cb_admin_token` | ` { query: { feedbackScore }` | `[ { result: { cbEmail, feedbackScore, feedback_date } } ]`| ✨ shiny shiny new ✨ |
```
Comment (eliascodes) [inline docs/api_migration_plan.md]: This is still inaccurate:

screen shot 2018-03-15 at 10 57 00

File: react-backend/client/src/components/visitors/thank_you_feedback.js (hunk 26 above)
Comment (astroash) [reply react-backend/client/src/components/visitors/thank_you_feedback.js]: I was contemplating passing a key to `props` for the component to match on to add the correct copy, but I felt like this wasn't simplifying the already simplified functional component much and would take away from readability, so settled for this 🤷🏽‍♀️. I mean, if it was elm I would have union typed the crap out of it 😏. Happy to implement the above, or any other suggestion if ya want.

Comment (eliascodes) [reply react-backend/client/src/components/visitors/thank_you_feedback.js]: Nah leave it, just fix the docs.
